File: resources/hooks/CodePulseHook.py
import argparse
import hashlib
import hmac
import json
import os
import sys
import time
import uuid

MAX_INPUT_LENGTH = 2 * 1024 * 1024


def main() -> int:
    parser = argparse.ArgumentParser(add_help=False)
    parser.add_argument("--token", required=True)
    parser.add_argument("--inbox", required=True)
    parser.add_argument("--distro", default="")
    args = parser.parse_args()
    event_name = ""
    try:
        if len(args.token) != 64 or any(value not in "0123456789abcdefABCDEF" for value in args.token):
            return 0
        raw = sys.stdin.buffer.read(MAX_INPUT_LENGTH + 1)
        if not raw or len(raw) > MAX_INPUT_LENGTH:
            return 0
        source = json.loads(raw)
        event_name = source.get("hook_event_name", "")
        session_id = source.get("session_id", "")
        if not isinstance(event_name, str) or not isinstance(session_id, str) or not event_name or not session_id:
            return 0
        payload = {
            "hook_event_name": event_name,
            "session_id": session_id,
            "cwd": source.get("cwd", "") if isinstance(source.get("cwd", ""), str) else "",
            "tool_name": source.get("tool_name", "") if isinstance(source.get("tool_name", ""), str) else "",
            "runtime": "wsl",
            "distro": args.distro,
        }
        body = json.dumps(payload, ensure_ascii=False, separators=(",", ":"))
        signature = hmac.new(bytes.fromhex(args.token), body.encode("utf-8"), hashlib.sha256).hexdigest()
        os.makedirs(args.inbox, exist_ok=True)
        name = f"{time.time_ns()}-{os.getpid()}-{uuid.uuid4().hex}.json"
        temporary = os.path.join(args.inbox, f".{name}.tmp")
        target = os.path.join(args.inbox, name)
        with open(temporary, "x", encoding="utf-8") as output:
            json.dump({"body": body, "signature": signature}, output, ensure_ascii=False, separators=(",", ":"))
        os.replace(temporary, target)
    except Exception:
        pass
    finally:
        if event_name in ("Stop", "SubagentStop"):
            sys.stdout.write("{}")
            sys.stdout.flush()
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: resources/hooks/CodePulseHook.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Net;
using System.Text;
using System.Web.Script.Serialization;

internal static class CodePulseHook
{
    private const int MaxInputLength = 2 * 1024 * 1024;

    [STAThread]
    private static int Main(string[] args)
    {
        string eventName = null;
        try
        {
            Forward(args, out eventName);
        }
        catch (Exception error)
        {
            // Monitoring must never interrupt a Codex task.
            if (Array.IndexOf(args, "--diagnostic") >= 0) Console.Error.WriteLine(error);
        }

        WriteHookOutput(eventName);

        return 0;
    }

    private static void Forward(string[] args, out string eventName)
    {
        eventName = null;
        string token = ArgumentValue(args, "--token");
        if (!IsToken(token)) return;

        string raw = ReadStandardInput();
        if (String.IsNullOrWhiteSpace(raw)) return;

        JavaScriptSerializer serializer = new JavaScriptSerializer { MaxJsonLength = MaxInputLength };
        Dictionary<string, object> input = serializer.Deserialize<Dictionary<string, object>>(raw);
        if (input == null) return;

        eventName = StringValue(input, "hook_event_name");
        string sessionId = StringValue(input, "session_id");
        if (String.IsNullOrEmpty(eventName) || String.IsNullOrEmpty(sessionId)) return;

        Dictionary<string, object> payload = new Dictionary<string, object>
        {
            { "hook_event_name", eventName },
            { "session_id", sessionId },
            { "cwd", StringValue(input, "cwd") },
            { "tool_name", StringValue(input, "tool_name") },
            { "runtime", ArgumentValue(args, "--runtime") == "wsl" ? "wsl" : "windows" },
            { "distro", ArgumentValue(args, "--distro") ?? String.Empty }
        };

        int port = 17322;
        int diagnosticPort;
        if (Array.IndexOf(args, "--diagnostic") >= 0 &&
            Int32.TryParse(ArgumentValue(args, "--port"), out diagnosticPort) &&
            diagnosticPort > 0 && diagnosticPort <= 65535)
        {
            port = diagnosticPort;
        }

        byte[] body = Encoding.UTF8.GetBytes(serializer.Serialize(payload));
        HttpWebRequest request = (HttpWebRequest)WebRequest.Create("http://127.0.0.1:" + port + "/activity");
        request.Method = "POST";
        request.ContentType = "application/json";
        request.ContentLength = body.Length;
        request.Headers[HttpRequestHeader.Authorization] = "Bearer " + token;
        request.KeepAlive = false;
        request.Proxy = null;
        request.Timeout = 1200;
        request.ReadWriteTimeout = 1200;

        using (Stream stream = request.GetRequestStream())
        {
            stream.Write(body, 0, body.Length);
        }

        using (HttpWebResponse response = (HttpWebResponse)request.GetResponse())
        {
            // Waiting for the response guarantees that CodePulse accepted the event.
        }
    }

    private static void WriteHookOutput(string eventName)
    {
        if (eventName != "Stop" && eventName != "SubagentStop") return;
        using (StreamWriter writer = new StreamWriter(Console.OpenStandardOutput(), new UTF8Encoding(false)))
        {
            writer.Write("{}");
            writer.Flush();
        }
    }

    private static string ReadStandardInput()
    {
        using (StreamReader reader = new StreamReader(Console.OpenStandardInput(), Encoding.UTF8, true, 4096))
        {
            char[] buffer = new char[4096];
            StringBuilder value = new StringBuilder();
            int count;
            while ((count = reader.Read(buffer, 0, buffer.Length)) > 0)
            {
                if (value.Length + count > MaxInputLength) return String.Empty;
                value.Append(buffer, 0, count);
            }
            return value.ToString();
        }
    }

    private static string ArgumentValue(string[] args, string name)
    {
        for (int index = 0; index + 1 < args.Length; index++)
        {
            if (args[index] == name) return args[index + 1];
        }
        return null;
    }

    private static string StringValue(Dictionary<string, object> input, string name)
    {
        object value;
        return input.TryGetValue(name, out value) && value is string ? (string)value : String.Empty;
    }

    private static bool IsToken(string token)
    {
        if (String.IsNullOrEmpty(token) || token.Length != 64) return false;
        foreach (char value in token)
        {
            bool hexadecimal = (value >= '0' && value <= '9') ||
                               (value >= 'a' && value <= 'f') ||
                               (value >= 'A' && value <= 'F');
            if (!hexadecimal) return false;
        }
        return true;
    }

}
